docs/src/docs/asciidoc/developer/clustered-events.adoc

```diff
@@ -0,0 +1,90 @@
+a very noticeable one.
+
+Some features are undesirable because they are unlikely to be practical:
+
+* Synchronous events would require waiting for a round-trip to all clients before achieving a cache operation. This
+would pretty much make such cache unusably slow. Such config should throw *UnsupportedOperationException* when attempted.
+* Guaranteeing event delivery in all cases would require some form of stable store and a fairly complex and costly
+2-phase logic. This would also have an unsustainable performance impact. Instead, clients should have a way to figure out
+when such hazard happen to compensate for the possible lack of event delivery.
+
+== Technical facts
+
+Because of the current implementation:
+
+* Most events have to find their source from a client, more specifically from the clustered `Store` implementation:
+*expiry*, *removal*, *update* and *creation* are all client-triggered.
+* One event has to originate from the server: *eviction* is server-triggered.
+
+This means a cluster-wide listener mechanism has to be created with the following features:
+
+* Clients can register and unregister themselves.
+* Events can be fired from any client or any server.
+* Events have to be transported either from a server to all listening clients, or from one client to all clients
+(sometimes including the source client) via a server. AA topologies complicate client-sourced events transport as events
+have to be delivered only once.
+* Events can be fired from non-listening clients.
+
+== The transport mechanism
+
+An event delivery mechanism must be built to transport the events from the clients and servers to all clients registered
+as listeners. It requires the following:
+
+* API to (un)register a client as a listener
+* API for a client to fire an event
+* API for a server to fire an event
+* Implementation support of multiple servers
+
+A dedicated entity looks like the cleanest way ahead. This entity might need to be general-purpose enough to make
+it able to also transport invalidation events, and possibly any other type of notification logic between clients.
+Then refactor the invalidation mechanism to make use of this new notification entity. This should be regarded as a
+stretch goal once the implementation of this story is done.
+
+== The straightforward bits
+
+Assuming the event delivery mechanism is in place event firing logic for the following `Store` methods is going to be
+straightforward to implement as all the necessary data to decide what event to fire already is available handy:
+
+* putIfAbsent()
+* remove(1)
+* remove(2)
+* replace(2)
+* replace(3)
+
+The `Store` methods not listed above or below either throw `UnsupportedOperationException`, or delegate to other
+`Store` methods.
+
+== The complicated bits
+
+The following cases are going to be more complicated to implement:
+
+* All clients must be aware of the event firing logic, apply its rules and fire events even when some clients are not
```

Comment:
Wouldn't this whole thing be simpler if we layered the event listener feature in the same way that the current clustered caching is layered. That would mean building a "chain event listener" feature that the server implements, and then have the clients build their feature on top.  In this model events are only sourced from the server... there is a slight complication around "expiry" but that complication already exists. The only thing this would cause us to do is to clarify when an expiration event occured, with the constraints here forcing that to be "when a client successfully resolves a value to expired on the server". This may require associating some additional metadata with the resolve operation but I don't think that will prove insurmountable.

Reply:
If I get you idea right, the server currently lacks data that would allow the clients to figure out what mapping was modified.